### DISCUSSION TITLE
Nand2Tetris 소개

### DISCUSSION THREAD

**Discussion by obzva** [DaleStudy/leetcode-study#514] [자료공유]
이번엔 캐주얼하게 자료 소개 하나 남겨볼까 합니다.

#284 문제처럼 `+`, `-` 연산자를 사용하지 않고 `+`, `-` 기능을 수행하게끔 하는 문제는 보통 논리 연산자와 bitwise 연산자를 이용하여 풀이합니다.

많은 분들이 web 프론트/백엔드 개발에 종사하시기에 논리 연산자와 bitwise가 실무에서 쓰이는 일은 극히 적을 것으로 예상합니다..

만, `덧셈 뺄셈이 컴퓨터/하드웨어에서 어떻게 구현되어 있는지 궁금하다`거나 `논리 연산자를 이용해서 이것 저것 해보는게 재밌어 보인다`는 생각이 든다면 아래 강의를 추천 드립니다.

- coursera 강의 (무료): [Nand2Tetris](https://www.coursera.org/learn/build-a-computer)
- 책으로는 [밑바닥부터 만드는 컴퓨팅 시스템](https://m.yes24.com/Goods/Detail/71129079)이 위 강의 내용을 다루고 있습니다

강의 및 책의 내용을 짧게 요약하자면, `NAND` 게이트로 간단한 연산 칩을 구성해보는 것부터 시작하여 앞서 만든 칩들로 컴퓨터를 만들어본 뒤에 컴파일러와 JVM과 비슷한 가상 머신까지 직접 만들어보는 실습 교본입니다.

개인적으로는 아래와 같은 점들이 도움된다고 느꼈습니다 
- 컴퓨터 구조의 대략적인 틀을 머릿속에 그리는데 도움이 됨 (실습을 해보니까)
- 추상화 및 모듈 구성에 대한 기초적인 연습 및 복습
- (그냥 재밌음)

[OSSU](https://github.com/ossu/computer-science)의 `core systems` 커리큘럼의 첫번째 강좌로 채택되어 있기 때문에 혼자 공부하는 것이 힘들다면 `OSSU` 커뮤니티 (디스코드) 에서 질문을 주고받는 것을 추천드리며,  [Nand2Tetris QnA Forum](http://nand2tetris-questions-and-answers-forum.52.s1.nabble.com/Hardware-f32586.html) 또한 참고하실 수 있습니다.

![image](https://github.com/user-attachments/assets/b9cea4d2-5b96-4e07-8f7b-539b6a5857c4)

**Comment by SamTheKorean** [comment]
이 책 실습은 안하고 절반 정도만 읽었는데 마무리해야겠습이다! ㅎㅎ

**Reply by obzva** [reply]
눈으로 봤을 땐 간단해 보이는데 실습해보면 좀 까다로운 과제가 좀 있더라구요 ㅎㅎ

**Comment by DaleSeo** [comment]
> (그냥 재밌음)

거리감 느껴지게 하지 마세요 ㅋㅋ

**Comment by bky373** [comment]
오.. 흥미로운 내용이군요! 좋은 자료 추천해주셔서 감사합니다! 👍 

> 거리감 느껴지게 하지 마세요 ㅋㅋ

ㅋㅋㅋㅋㅋㅋ 저도 살짝 거리감이..
